precompiles/assets/types.go

```diff
@@ -32,27 +36,121 @@ func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []i
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, reflect.TypeOf(args[0]), args[0])
+	}
+	clientChain.LayerZeroChainID = uint64(clientChainID)
+
+	addressLength, ok := args[1].(uint32)
+	if !ok || addressLength == 0 {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, reflect.TypeOf(args[1]), args[1])
+	}
+	if addressLength > types.GeneralAssetsAddrLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, addressLength, "not greater than 32")
+	}
+	clientChain.AddressLength = addressLength
+
+	name, ok := args[2].(string)
+	if !ok || name == "" {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, reflect.TypeOf(args[2]), args[2])
+	}
+	clientChain.Name = name
+
+	metaInfo, ok := args[3].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, reflect.TypeOf(args[2]), args[2])
+	}
+	clientChain.MetaInfo = metaInfo
+
+	signatureType, ok := args[4].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 4, reflect.TypeOf(args[4]), args[4])
+	}
+	clientChain.SignatureType = signatureType
+
+	return &clientChain, nil
+}
+
+func (p Precompile) TokensFromInputs(ctx sdk.Context, args []interface{}) ([]types.AssetInfo, error) {
+	if len(args) != 4 {
+		return nil, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 4, len(args))
+	}
+	assets := make([]types.AssetInfo, 0)
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, reflect.TypeOf(args[0]), args[0])
+	}
+	info, err := p.assetsKeeper.GetClientChainInfoByIndex(ctx, uint64(clientChainID))
+	if err != nil {
+		return nil, err
+	}
+	clientChainAddrLength := info.AddressLength
+
+	assetAddrList, ok := args[1].([][]byte)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, reflect.TypeOf(args[1]), args[1])
+	}
+	assetNumber := len(assetAddrList)
+	if assetNumber == 0 {
+		return nil, fmt.Errorf(exocmn.ErrInvalidInputList, len(assetAddrList), "greater than 0")
+	}
+	decimalList, ok := args[2].([]uint8)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 2, reflect.TypeOf(args[2]), args[2])
+	}
+	if assetNumber != len(decimalList) {
+		return nil, fmt.Errorf(exocmn.ErrInvalidInputList, len(decimalList), assetNumber)
+	}
+	tvlLimitList, ok := args[3].([]*big.Int)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 3, reflect.TypeOf(args[3]), args[3])
+	}
+	if assetNumber != len(tvlLimitList) {
+		return nil, fmt.Errorf(exocmn.ErrInvalidInputList, len(tvlLimitList), assetNumber)
+	}
+
+	for i := 0; i < assetNumber; i++ {
+		if len(assetAddrList[i]) != types.GeneralAssetsAddrLength {
+			return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, len(assetAddrList[i]), types.GeneralClientChainAddrLength)
+		}
+		assetAddr := assetAddrList[i][:clientChainAddrLength]
+		assets = append(assets, types.AssetInfo{
+			Address:          hexutil.Encode(assetAddr),
```

Comment:
Add TODO for non-EVM chains

Reply:
It appears that the compatibility of non-EVM chains will be addressed in the Exocore gateway contract. However, it's still pending, so I will add a code comment.